prov/util/src/util_srx.c

```diff
@@ -0,0 +1,1187 @@
+static int util_srx_setopt(fid_t fid, int level, int optname,
+		           const void *optval, size_t optlen)
+{
+	struct util_srx_ctx *srx =
+		container_of(fid, struct util_srx_ctx, peer_srx.ep_fid.fid);
+
+	if ((level != FI_OPT_ENDPOINT) || (optname != FI_OPT_MIN_MULTI_RECV))
+		return -FI_ENOPROTOOPT;
+
+	srx->min_multi_recv_size = *(size_t *)optval;
+
+	return FI_SUCCESS;
+}
+
+static struct fi_ops_ep util_srx_ops = {
+	.size = sizeof(struct fi_ops_ep),
+	.cancel = util_srx_cancel,
+	.getopt = util_srx_getopt,
+	.setopt = util_srx_setopt,
+	.tx_ctx = fi_no_tx_ctx,
+	.rx_ctx = fi_no_rx_ctx,
+	.rx_size_left = fi_no_rx_size_left,
+	.tx_size_left = fi_no_tx_size_left,
+};
+
+static void util_rx_entry_init(struct ofi_bufpool_region *region, void *buf)
+{
+	struct util_rx_entry *rx_entry = (struct util_rx_entry *) buf;
+	struct util_srx_ctx *srx = (struct util_srx_ctx *)
+					region->pool->attr.context;
+
+	rx_entry->peer_entry.iov = util_srx_iov(rx_entry);
+	rx_entry->peer_entry.desc = util_srx_desc(srx, rx_entry);
+}
+
+static void util_srx_init_slist(struct ofi_dyn_arr *arr, void *item)
+{
+	slist_init((struct slist *) item);
+}
+
+static void util_srx_init_unexp_peer(struct ofi_dyn_arr *arr, void *item)
+{
+	struct util_unexp_peer *unexp_peer = item;
+
+	slist_init(&unexp_peer->msg_queue);
+	slist_init(&unexp_peer->tag_queue);
+	unexp_peer->cnt = 0;
+}
+
+int util_ep_srx_context(struct util_domain *domain, size_t rx_size,
+			size_t iov_limit, size_t default_min_multi_recv,
+			ofi_update_func_t update_func,
+			struct ofi_genlock *lock, struct fid_ep **rx_ep)
+{
+	struct util_srx_ctx *srx;
+	struct ofi_bufpool_attr pool_attr = {0};
+	int ret = FI_SUCCESS;
+
+	srx = calloc(1, sizeof(*srx));
+	if (!srx)
+		return -FI_ENOMEM;
+
+	ofi_array_init(&srx->src_unexp_peers, sizeof(struct util_unexp_peer),
+		       util_srx_init_unexp_peer);
+	dlist_init(&srx->unspec_unexp_msg_queue);
+	dlist_init(&srx->unspec_unexp_tag_queue);
+	dlist_init(&srx->unexp_peers);
+
+	ofi_array_init(&srx->src_recv_queues, sizeof(struct slist),
+		       util_srx_init_slist);
+	ofi_array_init(&srx->src_trecv_queues, sizeof(struct slist),
+		       util_srx_init_slist);
+
+	slist_init(&srx->msg_queue);
+	slist_init(&srx->tag_queue);
+
+	//each entry has the iovs and descriptors stored at the end of the entry
+	//calculate how much space each entry needs based on provider iov limits
+	pool_attr.size = sizeof(struct util_rx_entry) +
+		(sizeof(struct iovec) + sizeof(void *)) * iov_limit;
+	pool_attr.alignment = 16;
+	pool_attr.chunk_cnt = rx_size,
+	pool_attr.alloc_fn = NULL;
+	pool_attr.free_fn = NULL;
+	pool_attr.init_fn = util_rx_entry_init;
+	pool_attr.context = srx;
+	ret = ofi_bufpool_create_attr(&pool_attr, &srx->rx_pool);
+	if (ret) {
+		free(srx);
+		return ret;
+	}
+
+	srx->min_multi_recv_size = default_min_multi_recv;
+	srx->iov_limit = iov_limit;
+	srx->dir_recv = domain->info_domain_caps & FI_DIRECTED_RECV;
+	srx->update_func = update_func;
+	srx->lock = lock;
```

Comment:
Can we have a default lock here? So provider doesn't have to create itself?

Reply:
I originally had that but it made cleanup messy (since we can only destroy the lock if we created it). Would you like me to add that in for this PR?